### PR TITLE
fix: defineComponent() with array props

### PR DESCRIPTION
**Pull request by antfu**
This continues #150, as it's kinda out-of-date. The type declarations are ported from vue-next.

Before
image

After
image

### CODE REVIEW

**Comment by antfu** [comment]
Tests passed on my machine but failed on Github Actions. I guess it's because it uses npm without package-lock.json which may cause dependencies inconsistency. Hold to see if the error still persists after #347 gets merged.

**Comment by antfu** [comment]
Passed with ts `v3.9.4` but failed with the latest `v3.9.5`. Let's hold it for a while.

**Comment by pikax** [comment]
Updated to `3.9.5` on `vue-next` and no error, do you want me to fix this?

**Comment by antfu** [comment]
@pikax it's might due to that I have moved the array overload in front of the object one. And yes, it would be great if you can help out :)

**Comment by pikax** [comment]
Replace `PropOptions` with: 
```ts
type DefaultFactory<T> = () => T | null | undefined;

interface PropOptions<T = any> {
  type?: PropType<T> | true | null;
  required?: boolean;
  default?: T | DefaultFactory<T> | null | undefined;
  validator?(value: unknown): boolean;
}
```

File: test/types/defineComponent.spec.ts [hunk 1]
```diff
@@ -144,6 +144,17 @@ describe('defineComponent', () => {
     expect.assertions(2);
   });
 
+  it('should accept tuple props', () => {
+    const App = createComponent({
```
Comment (pikax) [inline test/types/defineComponent.spec.ts:148]: Do you mean `defineComponemt`?

Comment (antfu) [reply test/types/defineComponent.spec.ts]: Yes, just fixed. I also updated the type of createComponent

**Comment by antfu** [comment]
It works. Thanks!

**Comment by FloryanFilip** [comment]
Can I somehow define my prop as tuple value?
`			plannerDateRange: { type: Array as PropType<[string, string]>, required: true },
`
The above produces compilation error: 
Screenshot 2021-01-19 at 14 29 28

**Pull request merged**